### PR TITLE
Alternative workaround for DMD 14447 in std.utf.decodeImpl()

### PR DESCRIPTION
**Pull request by tsbockman**
This one is more complicated than [my other](https://github.com/D-Programming-Language/phobos/pull/4072), but also fixes an issue with improper use of `@trusted`. The high line count of the diff is mostly because I had to change the indentation level for the bodies of `decodeImpl()`.

The character array specializations of `decodeImpl()` use some pointer math, which should nevertheless be safe. So, those instantiations are supposed to be `@trusted`. Currently, *callers* of `decodeImpl()` are being marked as `@trusted`, whereas they should be `@safe`.

This PR fixes that, as well as implementing a better workaround for [DMD issue #14447](https://issues.dlang.org/show_bug.cgi?id=14447), allowing me to eliminate the redundant `canIndex` template parameter of `decodeImpl()`.

See also:
* My [simpler alternative](https://github.com/D-Programming-Language/phobos/pull/4072) to this PR, which does not fix the `@trusted` problem.
* @jmdavis 's PR which [further reworks](https://github.com/D-Programming-Language/phobos/pull/4012) this area of `std.utf`

### CODE REVIEW

**Comment by jmdavis** [comment]
I'll try and look at this later today or tomorrow, but please remove the changes to `UseReplacementDchar` from this PR. Whether we should make that change is a completely separate issue from the improvements to `decodeImpl`, and it's a discussion unto itself.

**Comment by tsbockman** [comment]
> please remove the changes to `UseReplacementDchar` from this PR.

OK. (Just figured I'd see if I could sneak that part through with this, since I don't care enough to make a separate PR for it.)

**Comment by DmitryOlshansky** [comment]
Ping @jmdavis

**Comment by DmitryOlshansky** [comment]
Makes sense to me

File: std/utf.d [hunk 1]
```diff
@@ -1005,8 +1005,7 @@ alias UseReplacementDchar = Flag!"useReplacementDchar";
         sequence and useReplacementDchar is UseReplacementDchar.no
   +/
 dchar decode(UseReplacementDchar useReplacementDchar = UseReplacementDchar.no, S)(auto ref S str, ref size_t index)
-    if (!isSomeString!S &&
-        isRandomAccessRange!S && hasSlicing!S && hasLength!S && isSomeChar!(ElementType!S))
+    if (isSomeString!S || isRandomAccessRange!S && hasSlicing!S && hasLength!S && isSomeChar!(ElementType!S))
```
Comment (wilzbach) [inline std/utf.d]: The other template constraint explicitly excludes strings - now they are allowed. Is this on purpose?

Comment (tsbockman) [inline std/utf.d]: I combined the range overload with the `if (isSomeString!S)` one directly below.

The only reason they were separate previously, was to make the `isSomeString` version `@trusted`. But my other changes make that unnecessary now.

**Comment by tsbockman** [comment]
Rebased.

**Comment by tsbockman** [comment]
Ping @jmdavis

**Comment by DmitryOlshansky** [comment]
Repinging @jmdavis

**Comment by tsbockman** [comment]
@jmdavis @DmitryOlshansky Rebased. Please review or reject.

Otherwise, I'm probably just going to close this next time it needs a rebase; it's too much trouble because git doesn't understand the structure of my changes and exaggerates the diff,.

**Comment by tsbockman** [comment]
This is not worth rebasing indefinitely. Closing.

**Comment by jmdavis** [comment]
@tsbockman Sorry about that. I haven't managed to spend much time on D PRs in recent months, and this fell by the wayside as a result, which it shouldn't have. I definitely dropped the ball on this one.

**Comment by tsbockman** [comment]
@jmdavis Honestly, this PR isn't important; it's just a bit of minor refactoring. If you're too busy to keep up with everything, this is probably a good one to ignore.

**Pull request closed without merging**